io-engine/src/bdev/ftl.rs

```diff
@@ -0,0 +1,258 @@
+
+    fn try_from(uri: &Url) -> Result<Self, Self::Error> {
+        let segments = uri::segments(uri);
+        if segments.is_empty() {
+            return Err(BdevError::InvalidUri {
+                uri: uri.to_string(),
+                message: "empty path".to_string(),
+            });
+        }
+
+        let mut parameters: HashMap<String, String> =
+            uri.query_pairs().into_owned().collect();
+
+        let uuid = uri::uuid(parameters.remove("uuid")).context(
+            bdev_api::UuidParamParseFailed {
+                uri: uri.to_string(),
+            },
+        )?;
+
+        let bbdev_uri = parameters
+            .remove("bbdev")
+            .context(bdev_api::InvalidUri {
+                uri: uri.to_string(),
+                message: String::from("No bbdev parameter found"),
+            })?
+            .to_string()
+            .replace("!", "?")
+            .replace("|", "&");
+
+        let cbdev_uri = parameters
+            .remove("cbdev")
+            .context(bdev_api::InvalidUri {
+                uri: uri.to_string(),
+                message: String::from("No cbdev parameter found"),
+            })?
+            .to_string()
+            .replace("!", "?")
+            .replace("|", "&");
+
+        // Device parameter checking for bbdev and cbdev are done in bdev_ftl_create_bdev
+        reject_unknown_parameters(uri, parameters)?;
+
+        Ok(Self {
+            name: uri.path()[1..].into(),
+            alias: uri.to_string(),
+            uuid,
+            bbdev_uri,
+            cbdev_uri,
+        })
+    }
+}
+
+impl GetName for Ftl {
+    fn get_name(&self) -> String {
+        self.name.clone()
+    }
+}
+
+pub extern "C" fn ftl_bdev_init_fn_cb(
+    _ptr: *const ftl_bdev_info,
+    sender_ptr: *mut c_void,
+    errno: i32,
+) {
+    info!("{:?}: ftl_bdev_init_fn_cb", errno);
+    let sender = unsafe {
+        Box::from_raw(sender_ptr as *mut oneshot::Sender<ErrnoResult<()>>)
+    };
+    sender
+        .send(errno_result_from_i32((), errno))
+        .expect("done callback receiver side disappeared");
+}
+
+#[async_trait(?Send)]
+impl CreateDestroy for Ftl {
+    type Error = BdevError;
+
+    async fn create(&self) -> Result<String, Self::Error> {
+        if UntypedBdev::lookup_by_name(&self.name).is_some() {
+            return Err(BdevError::BdevExists {
+                name: self.name.clone(),
+            });
+        }
+
+        debug!(
+            "{:?}: Creating ftl-bdev '{:?}' with bbdev {:?} and cbdev {:?}.",
+            self,
+            CString::new(self.name.clone()).unwrap(),
+            self.bbdev_uri,
+            self.cbdev_uri
+        );
+
+        use std::ffi::CString;
+        extern crate libc;
+
+        let base_dev_name = bdev_create(&self.bbdev_uri).await?;
+        let cache_dev_name = bdev_create(&self.cbdev_uri).await?;
+
+        let l_name_ptr =
+            CString::new(self.name.clone()).unwrap().into_raw() as *mut i8;
```

Comment:
This is leaking memory?
Does it have to be mut or can we use a reference? example:
```suggestion
        let l_name =
            CString::new(self.name.clone()).unwrap();
```
```
 ftl_conf.name = l_name.as_ptr();
```

Reply:
Good catch, yes it is leaking memory! I reworked this section.